File: src/main/java/noteedit/service/FolderService.java
package com.gliese.noteedit.service;

import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl;
import com.gliese.noteedit.entity.FileInfo;
import com.gliese.noteedit.entity.FolderInfo;
import com.gliese.noteedit.mapper.FolderMapper;
import com.gliese.noteedit.utils.BasicUtils;
import com.gliese.noteedit.utils.FileUtils;
import com.gliese.noteedit.utils.JsonUtils;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.*;

@Slf4j
@Service
public class FolderService extends ServiceImpl<FolderMapper, FolderInfo> {
    @Autowired
    private FolderMapper folderMapper;

    /**
     * 在数据库中创建新文件夹
     * @param fileInfo 文件信息
     * @return 文件所在目录id
     */
    public Long mkdirs(FileInfo fileInfo) {
        Long num = folderMapper.selectCount(new QueryWrapper<>());
        QueryWrapper<FolderInfo> queryWrapper = new QueryWrapper<>();
        queryWrapper.orderByAsc("depth").in("folder_name", fileInfo.getFolders());
        List<FolderInfo> existsFolders = folderMapper.selectList(queryWrapper);
        for(FolderInfo folder: existsFolders) {
            List<Integer> children = (List<Integer>) JsonUtils.jsonToObject(folder.getChildrenJson(), List.class);
            List<Long> children2 = new ArrayList<>();
            for(Integer child: children) {
                children2.add(child.longValue());
            }
            folder.setChildren(children2);
        }
        List<FolderInfo> newFolders = FileUtils.makeNewFolders(existsFolders, fileInfo.getFolders(), num);
        FolderInfo updateFolder = newFolders.get(newFolders.size() - 1);
        newFolders.remove(newFolders.size() - 1);
        if(updateFolder != null)
            folderMapper.updateById(updateFolder);
        this.saveOrUpdateBatch(newFolders, 100);
        if(newFolders.size() > 0) { // * 有新目录
            log.info("new folders made");
            return newFolders.get(newFolders.size() - 1).getId();
        }
        log.info("no new folders found");
        return updateFolder.getId(); // * 无新目录
    }

    /**
     * 获取根目录
     */
    public List<FolderInfo> listRoot() {
        QueryWrapper<FolderInfo> queryWrapper = new QueryWrapper<>();
        queryWrapper.eq("depth", 1);
        List<FolderInfo> result = folderMapper.selectList(queryWrapper);
        return result;
    }

    public FolderInfo getFolderById(Long id) {
        if(id == null)
            return null;
        QueryWrapper<FolderInfo> queryWrapper = new QueryWrapper<>();
        queryWrapper.eq("id", id);
        FolderInfo result = folderMapper.selectOne(queryWrapper);
        result.setChildren(BasicUtils.intToLongBatch((List<Integer>) JsonUtils.jsonToObject(result.getChildrenJson(), List.class)));
        return result;
    }
    public List<FolderInfo> getFoldersById(List<Long> ids) {
        if(ids == null || ids.size() == 0)
            return null;
        if(ids.size() == 1) {
            List<FolderInfo> result = new ArrayList<>();
            result.add(getFolderById(ids.get(0)));
            return result;
        }
        QueryWrapper<FolderInfo> queryWrapper = new QueryWrapper<>();
        queryWrapper.in("id", ids);
        List<FolderInfo> results = folderMapper.selectList(queryWrapper);
        if(results == null)
            return null;
        for(FolderInfo result: results) {
            result.setChildren(BasicUtils.intToLongBatch((List<Integer>) JsonUtils.jsonToObject(result.getChildrenJson(), List.class)));
        }
        return results;
    }

    public FolderInfo getLeafFolderByPath(List<String> path) {
        QueryWrapper<FolderInfo> queryWrapper = new QueryWrapper<>();
        queryWrapper.in("folder_name", path);
        List<FolderInfo> results = folderMapper.selectList(queryWrapper);
        Map<Long, Integer> map = new HashMap<>();
        Queue<Integer> q = new LinkedList<>();
        int index = 0;
        for(FolderInfo result: results) {
            map.put(result.getId(), index++);
            result.setChildren(BasicUtils.intToLongBatch((List<Integer>) JsonUtils.jsonToObject(result.getChildrenJson(), List.class)));
            if(result.getDepth() == 1)
                q.add(index - 1);
        }
        int depth = 0;
        FolderInfo leaf = null;
        while(!q.isEmpty()) {
            int top = q.poll();
            FolderInfo folder = results.get(top);
            if(folder.getFolderName().equals(path.get(depth))) {
                depth++;
                leaf = folder;
                for(Long childId: folder.getChildren())
                    if(map.containsKey(childId))
                        q.add(map.get(childId));
            }
        }
        return leaf;
    }
}

File: src/main/java/noteedit/service/FolderService.kt
package noteedit.service

import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper
import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl
import mu.KotlinLogging
import noteedit.entity.FolderInfo
import noteedit.exception.IllegalPathException
import noteedit.mapper.FolderMapper
import noteedit.utils.FileUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import javax.annotation.Resource

@Service
open class FolderService: ServiceImpl<FolderMapper, FolderInfo>() {
    private val logger = KotlinLogging.logger {}

    @Autowired
    private lateinit var folderMapper: FolderMapper

    /**
     * 新建文件夹
     * @param locatedFolders 文件所在文件夹名称
     * @return 文件所在目录id
     */
    open fun addNewList(locatedFolders: List<String>):Long {
        val tot = folderMapper.selectCount(null) ?: 0
        return (folderMapper.selectList(
            QueryWrapper<FolderInfo>()
                .orderByAsc("depth")
                .`in`("name", locatedFolders)
        ) ?: emptyList()).let { existFolders ->
            FileUtils.listNewFolders(existFolders, locatedFolders, tot)
        }.also { newFolders ->
            this.saveOrUpdateBatch(newFolders, newFolders.size)
        }.let {
            if(it.size > 0) {
                logger.info { "added ${it.size - 1} folders, update 1 folder" }
                it.last().id
            } else {
                0
            }
        }
    }

    /**
     * 获取根目录
     * @return 根目录
     */
    open fun listRoot(): List<FolderInfo> {
        return folderMapper.selectList(
            QueryWrapper<FolderInfo>()
                .eq("depth", 1)
        ) ?: emptyList()
    }

    /**
     * 根据id获取文件夹
     * @param id 文件夹id
     * @return 文件夹
     */
    open fun getById(id: Long): FolderInfo? {
        return folderMapper.selectById(id)
    }

    /**
     * 根据ids批量获取文件夹
     * @param ids 文件夹id
     * @return 文件夹列表
     */
    open fun getListByIds(ids: MutableList<Long>): List<FolderInfo> {
        require(ids.isNotEmpty()) { "ids is empty" }
        return folderMapper.selectList(
            QueryWrapper<FolderInfo>()
                .`in`("id", ids)
        ) ?: emptyList()
    }

    /**
     * 根据路径获取末尾文件夹
     * @param parts 文件夹路径
     * @return 叶子文件夹
     * @throws IllegalPathException 路径错误
     */
    open fun getLeafFolderByPath(parts: List<String>): FolderInfo {
        val folders = folderMapper.selectList(
            QueryWrapper<FolderInfo>()
                .`in`("name", parts)
                .orderByAsc("depth")
        ) ?: emptyList()
        for(i in 1 until folders.size) {
            require(folders[i - 1].subfolders?.contains(folders[i].id) ?: false) {
                IllegalPathException("路径非链式")
            }
        }
        return folders.last()
    }
}